File: Rubberduck.Parsing/Grammar/VBABaseParser.cs
using Antlr4.Runtime;
using System;
using System.Text.RegularExpressions;

namespace Rubberduck.Parsing.Grammar
{
    public abstract class VBABaseParser : Parser
    {
        public VBABaseParser(ITokenStream input) : base(input) { }

        #region Semantic predicate helper methods
        protected int TokenTypeAtRelativePosition(int i)
        {
            return _input.La(i);
        }

        protected IToken TokenAtRelativePosition(int i)
        {
            return _input.Lt(i);
        }

        protected string TextOf(IToken token)
        {
            return token.Text;
        }

        protected bool MatchesRegex(string text, string pattern)
        {
            return Regex.Match(text,pattern).Success;
        }

        protected bool EqualsStringIgnoringCase(string actual, string expected)
        {
            return actual.Equals(expected,StringComparison.OrdinalIgnoreCase);
        }

        protected bool EqualsStringIgnoringCase(string actual, params string[] expectedOptions)
        {
            foreach (string expected in expectedOptions)
            {
                if (actual.Equals(expected,StringComparison.OrdinalIgnoreCase))
                {
                    return true;
                }
            }
            return false;
        }

        protected bool EqualsString(string actual, string expected)
        {
            return actual.Equals(expected,StringComparison.Ordinal);
        }

        protected bool EqualsString(string actual, params string[] expectedOptions)
        {
            foreach (string expected in expectedOptions)
            {
                if (actual.Equals(expected,StringComparison.Ordinal))
                {
                    return true;
                }
            }
            return false;
        }

        protected bool IsTokenType(int actual, params int[] expectedOptions)
        {
            foreach (int expected in expectedOptions)
            {
                if (actual == expected)
                {
                    return true;
                }
            }
            return false;
        }
        #endregion
    }
}

File: development/java/Rubberduck.Parsing/Grammar/src/main/java/com/rubberduckvba/rubberduck/parsing/grammar/VBABaseParser.java
package com.rubberduckvba.rubberduck.parsing.grammar;

import org.antlr.v4.runtime.Parser;
import org.antlr.v4.runtime.Token;
import org.antlr.v4.runtime.TokenStream;

import java.util.regex.Pattern;

public abstract class VBABaseParser extends Parser {
    public VBABaseParser(TokenStream input) {
        super(input);
    }

    protected final int TokenTypeAtRelativePosition(int i) {
        return this._input.LA(i);
    }

    protected final Token TokenAtRelativePosition(int i) {
        return this._input.LT(i);
    }
	
    protected final String TextOf(Token token) {
        return token.getText();
    }

    protected final boolean MatchesRegex(String text, String pattern) {
        return Pattern.compile(pattern).matcher(text).matches();
    }

    protected final boolean EqualsStringIgnoringCase(String actual, String expected) {
        return actual.equalsIgnoreCase(expected);
    }

    protected final boolean EqualsStringIgnoringCase(String actual, String... expectedOptions) {
        for (String expected : expectedOptions) {
            if (actual.equalsIgnoreCase(expected)) {
                return true;
            }
        }
        return false;
    }

    protected final boolean EqualsString(String actual, String expected) {
        return actual.equals(expected);
    }

    protected final boolean EqualsString(String actual, String... expectedOptions) {
        for (String expected : expectedOptions) {
            if (actual.equals(expected)) {
                return true;
            }
        }
        return false;
    }

    protected final boolean IsTokenType(int actual, int... expectedOptions) {
        for (int expected : expectedOptions) {
            if (actual == expected) {
                return true;
            }
        }
        return false;
    }
}
